File: ports/haskell/src/Unicode/Security/Identity/EmojiZwjIntegrity.hs
{-|
Module      : Unicode.Security.Identity.EmojiZwjIntegrity
Description : Detection of malformed / unsanctioned emoji ZWJ-sequence shapes.

Haskell port of @Unicode.Security.Identity.EmojiZwjIntegrity@ from
unicode-lean, transliterated byte-faithfully from the verified Rust reference
implementation (the identity-layer
detector I3).

Threat model. An adversary crafts an emoji-shaped codepoint sequence
containing one or more @U+200D@ ZERO WIDTH JOINERs but violating the
sanctioned RGI ZWJ-sequence shape — by exceeding the RGI length cap, by
joining a non-emoji codepoint, by emitting adjacent ZWJ pairs, or by
overflowing the skin-tone count. Any non-RGI ZWJ-containing sequence is
renderer-dependent, and that renderer divergence is the attack surface.

Sanctioning data. UTS #51 defines the RGI ZWJ sequences in
@emoji-zwj-sequences.txt@, bundled in the port's own
@data/emoji-zwj-sequences.txt@ (never a host emoji library) and parsed through
the same NOINLINE runtime-table idiom the rest of the security layer uses. The
registered set gives both the exact-match membership test
('isRegisteredZwjSequence') and the ZWJ /alphabet/ — every distinct codepoint
occurring at any position of any registered sequence, excluding the joiner —
which is the canonical "what may flank a ZWJ?" predicate ('isEmojiTarget').

Algorithm (one pass over @input@).

  * Phase 1 — collect ZWJ positions and the skin-tone count.
  * Phase 2 — short-circuit 'Clear' if there are no ZWJs and the skin-tone
    count is at most 1.
  * Phase 3 — a registered RGI sequence is always 'Clear'.
  * Phase 4 — check sub-threats by priority:

      1. 'DoubleZwj'            — ZWJ-ZWJ adjacency.
      2. 'NonEmojiInjection'    — ZWJ adjacent to a non-emoji codepoint.
      3. 'OverLength'           — sequence longer than the RGI cap.
      4. 'SkinToneOverflow'     — skin-tone count @>= 5@.
      5. 'UnregisteredSequence' — catch-all when ZWJs are present but the
                                  sequence is not registered.
-}
module Unicode.Security.Identity.EmojiZwjIntegrity
  ( maxRgiLength
  , zwj
  , SubThreat
      ( DoubleZwj, NonEmojiInjection, OverLength, SkinToneOverflow
      , UnregisteredSequence
      )
  , subThreatTag
  , Classification (Clear, Hazard)
  , classificationIsClear
  , classificationTag
  , classificationPositions
  , Verdict
      ( Verdict, verdictInput, verdictClassify, verdictZwjPositions
      , verdictChainLength, verdictIsRegisteredRgi, verdictSkinToneCount
      )
  , isRegisteredZwjSequence
  , isEmojiTarget
  , isZwj
  , isEmojiModifier
  , detect
  , reasonCode
  ) where

import Data.Char (isSpace)
import Data.List (dropWhileEnd)
import Data.Maybe (listToMaybe, mapMaybe)
import Data.Set (Set)
import qualified Data.Set as Set
import Numeric (readHex)
import System.IO.Unsafe (unsafePerformIO)

import Paths_unicode_haskell (getDataFileName)

-- ─────────────────────────────────────────────────────────────────────
-- §1 Constants
-- ─────────────────────────────────────────────────────────────────────

-- | Conservative cap on the length of a sanctioned RGI ZWJ sequence
-- (@maxRgiLength@ in the Lean spec). The longest current entry (a four-person
-- family with skin tones) reaches ~13-14 codepoints; 16 is a safe upper bound.
maxRgiLength :: Int
maxRgiLength = 16

-- | The ZERO WIDTH JOINER codepoint.
zwj :: Int
zwj = 0x200D

-- ─────────────────────────────────────────────────────────────────────
-- §2 Types
-- ─────────────────────────────────────────────────────────────────────

-- | Sub-threats this detector can fire, in priority order. Each payload
-- carries the position values the conformance harness's attribution column
-- reads back, mirroring the Rust @SubThreat@ variants field-for-field.
data SubThreat
  = -- | ZWJ-ZWJ adjacency; the '[Int]' holds the first ZWJ of each adjacent
    -- pair.
    DoubleZwj [Int]
  | -- | A ZWJ flanked by a non-emoji codepoint (or sitting at an input edge).
    -- Fields: position of the offending ZWJ, then the non-emoji codepoint that
    -- flanks it (@0@ for an edge ZWJ).
    NonEmojiInjection Int Int
  | -- | The sequence is longer than 'maxRgiLength'. Fields: the observed
    -- sequence length, then the RGI length cap that was exceeded.
    OverLength Int Int
  | -- | Five or more skin-tone modifiers (the family-emoji maximum is four);
    -- the 'Int' is the observed skin-tone modifier count.
    SkinToneOverflow Int
  | -- | ZWJs are present and no other sub-threat matched, but the sequence is
    -- not a registered RGI ZWJ sequence; the 'Int' is the length of the
    -- unregistered ZWJ chain.
    UnregisteredSequence Int
  deriving stock (Eq, Show)

-- | Fixture-row tag string for this sub-threat (matches @SubThreat.tag@).
subThreatTag :: SubThreat -> String
subThreatTag (DoubleZwj _positions)              = "DoubleZWJ"
subThreatTag (NonEmojiInjection _zwjPos _cp)     = "NonEmojiInjection"
subThreatTag (OverLength _length _maxLength)     = "OverLength"
subThreatTag (SkinToneOverflow _count)           = "SkinToneOverflow"
subThreatTag (UnregisteredSequence _chainLen)    = "UnregisteredSequence"

-- | Top-level classification for EmojiZwjIntegrity.
data Classification
  = -- | A well-formed or non-ZWJ input.
    Clear
  | -- | A hazard: the fired sub-threat, the implicated codepoint positions,
    -- and the decoded-byte projection (always @[]@ here; kept for shape parity
    -- with the Lean @Classification.hazard@).
    Hazard SubThreat [Int] [Int]
  deriving stock (Eq, Show)

-- | True iff the classification is 'Clear'.
classificationIsClear :: Classification -> Bool
classificationIsClear Clear                = True
classificationIsClear (Hazard _sub _p _d)  = False

-- | Human-facing tag for a hazard, or 'Nothing' when clear.
classificationTag :: Classification -> Maybe String
classificationTag Clear                   = Nothing
classificationTag (Hazard sub _pos _dec)  = Just (subThreatTag sub)

-- | Implicated positions ('[]' when clear).
classificationPositions :: Classification -> [Int]
classificationPositions Clear                       = []
classificationPositions (Hazard _sub positions _d)  = positions

-- | Verdict — the structured output of 'detect' (mirrors the Lean @Verdict@).
data Verdict = Verdict
  { verdictInput           :: ![Int]
    -- ^ The scanned input codepoints.
  , verdictClassify        :: !Classification
    -- ^ The classification verdict.
  , verdictZwjPositions    :: ![Int]
    -- ^ Positions of every ZWJ in the input.
  , verdictChainLength     :: !Int
    -- ^ The chain length (@0@ when there are no ZWJs, else the input length).
  , verdictIsRegisteredRgi :: !Bool
    -- ^ True iff the input is exactly a registered RGI ZWJ sequence.
  , verdictSkinToneCount   :: !Int
    -- ^ Count of skin-tone modifier codepoints (@U+1F3FB@..@U+1F3FF@).
  }
  deriving stock (Eq, Show)

-- ─────────────────────────────────────────────────────────────────────
-- §3 RGI ZWJ-sequence data (bundled data/emoji-zwj-sequences.txt)
-- ─────────────────────────────────────────────────────────────────────

-- | The registered RGI ZWJ sequences parsed once from
-- @data/emoji-zwj-sequences.txt@ at first use, through the NOINLINE
-- runtime-table idiom the rest of the security layer uses.
zwjSequences :: [[Int]]
zwjSequences = unsafePerformIO $ do
  path <- getDataFileName "data/emoji-zwj-sequences.txt"
  parseZwjSequences <$> readFile path
{-# NOINLINE zwjSequences #-}

-- | Parse the registered RGI ZWJ sequences from @emoji-zwj-sequences.txt@.
-- Each non-comment row is @\<cp\> \<cp\> ... ; RGI_Emoji_ZWJ_Sequence ; \<desc\>@;
-- the codepoint list is the field before the first @;@.
parseZwjSequences :: String -> [[Int]]
parseZwjSequences = mapMaybe parseZwjLine . lines

-- | Parse one @emoji-zwj-sequences.txt@ row into its codepoint sequence, or
-- 'Nothing' when the row is blank, a comment, or carries a non-hex token (the
-- whole row is dropped on any parse failure, mirroring the Rust @break@).
parseZwjLine :: String -> Maybe [Int]
parseZwjLine raw =
  let body     = takeWhile (/= '#') raw
      stripped = trim body
  in if null stripped
       then Nothing
       else
         let seqField = takeWhile (/= ';') stripped
             tokens   = words seqField
         in case traverse parseHex tokens of
              Just cps | not (null cps) -> Just cps
              Just _empty               -> Nothing
              Nothing                   -> Nothing

-- | The ZWJ alphabet: every distinct codepoint occurring at any position of
-- any registered RGI ZWJ sequence, excluding the joiner @U+200D@ itself.
zwjAlphabet :: Set Int
zwjAlphabet =
  Set.fromList [ cp | codepoints <- zwjSequences, cp <- codepoints, cp /= zwj ]

-- | True iff @cps@ is exactly a registered RGI ZWJ sequence.
isRegisteredZwjSequence :: [Int] -> Bool
isRegisteredZwjSequence cps = any (== cps) zwjSequences

-- | True iff @cp@ appears at some position of a registered RGI ZWJ sequence
-- (the canonical "what may flank a ZWJ?" predicate).
isEmojiTarget :: Int -> Bool
isEmojiTarget cp = Set.member cp zwjAlphabet

-- ─────────────────────────────────────────────────────────────────────
-- §4 Core predicates
-- ─────────────────────────────────────────────────────────────────────

-- | True iff @cp@ is the ZWJ codepoint.
isZwj :: Int -> Bool
isZwj cp = cp == zwj

-- | True iff @cp@ is an emoji skin-tone modifier (@U+1F3FB@..@U+1F3FF@).
isEmojiModifier :: Int -> Bool
isEmojiModifier cp = 0x1F3FB <= cp && cp <= 0x1F3FF

-- | Positions of every ZWJ in @input@ (ascending).
zwjPositions :: [Int] -> [Int]
zwjPositions input = [ idx | (idx, cp) <- zip [0 ..] input, isZwj cp ]

-- | Count of skin-tone modifier codepoints.
skinToneCount :: [Int] -> Int
skinToneCount input = length (filter isEmojiModifier input)

-- | Positions of the first ZWJ in each ZWJ-ZWJ adjacent pair.
doubleZwjPositions :: [Int] -> [Int]
doubleZwjPositions input =
  [ idx
  | (idx, (cp, next)) <- zip [0 ..] (zip input (drop 1 input))
  , isZwj cp && isZwj next
  ]

-- | Codepoint at index @i@ of @input@, or 'Nothing' when @i@ is out of range.
-- Total; never uses the partial @!!@.
codepointAt :: [Int] -> Int -> Maybe Int
codepointAt input i
  | i < 0     = Nothing
  | otherwise = listToMaybe (drop i input)

-- | The first ZWJ position where either neighbour is a non-emoji codepoint, as
-- @(zwjPos, offendingCp)@. A ZWJ at an input edge (no preceding or no following
-- codepoint) is itself an injection-class hazard, reported with offending
-- codepoint @0@.
firstNonEmojiInjection :: [Int] -> Maybe (Int, Int)
firstNonEmojiInjection input = go 0 input
  where
    go :: Int -> [Int] -> Maybe (Int, Int)
    go _idx []           = Nothing
    go idx (cp : rest)
      | not (isZwj cp)   = go (idx + 1) rest
      | otherwise        =
          let prev = if idx == 0 then Nothing else codepointAt input (idx - 1)
              next = codepointAt input (idx + 1)
          in case (prev, next) of
               (Just prevCp, Just nextCp)
                 | not (isEmojiTarget prevCp) -> Just (idx, prevCp)
                 | not (isEmojiTarget nextCp) -> Just (idx, nextCp)
                 | otherwise                  -> go (idx + 1) rest
               (Nothing, _next)               -> Just (idx, 0)
               (Just _prevCp, Nothing)        -> Just (idx, 0)

-- ─────────────────────────────────────────────────────────────────────
-- §5 Top-level detection
-- ─────────────────────────────────────────────────────────────────────

-- | The EmojiZwjIntegrity detection function. See the module header for the
-- phase inventory and the sub-threat priority ladder.
detect :: [Int] -> Verdict
detect input =
  Verdict
    { verdictInput           = input
    , verdictClassify        = classification
    , verdictZwjPositions    = zwjs
    , verdictChainLength     = chainLen
    , verdictIsRegisteredRgi = isRgi
    , verdictSkinToneCount   = stCount
    }
  where
    zwjs     = zwjPositions input
    stCount  = skinToneCount input
    isRgi    = isRegisteredZwjSequence input
    chainLen = if null zwjs then 0 else length input

    classification
      -- Phase 2: no ZWJs and at most one skin tone is clear.
      | null zwjs && stCount <= 1 = Clear
      -- Phase 3: a registered RGI sequence is always clear.
      | isRgi                     = Clear
      | otherwise                 = phase4

    phase4 =
      let dzwj = doubleZwjPositions input
      in if not (null dzwj)
           -- Phase 4.1: ZWJ-ZWJ adjacency.
           then Hazard (DoubleZwj dzwj) dzwj []
           else case firstNonEmojiInjection input of
             -- Phase 4.2: ZWJ adjacent to a non-emoji codepoint.
             Just (zwjPos, offendCp) ->
               Hazard (NonEmojiInjection zwjPos offendCp) [zwjPos] []
             Nothing
               -- Phase 4.3: length cap.
               | length input > maxRgiLength ->
                   Hazard (OverLength (length input) maxRgiLength) [] []
               -- Phase 4.4: skin-tone overflow.
               | stCount >= 5 ->
                   Hazard (SkinToneOverflow stCount) [] []
               -- Phase 4.5: catch-all for unregistered ZWJ sequences.
               | not (null zwjs) ->
                   Hazard (UnregisteredSequence (length input)) zwjs []
               | otherwise -> Clear

-- | Fully-qualified reason code for a fired sub-threat, of the shape
-- @unicode.security.I.emoji-zwj-integrity.\<subThreatTag\>@. For
-- 'DoubleZwj' this is @unicode.security.I.emoji-zwj-integrity.DoubleZWJ@.
reasonCode :: SubThreat -> String
reasonCode sub = "unicode.security.I.emoji-zwj-integrity." ++ subThreatTag sub

-- ─────────────────────────────────────────────────────────────────────
-- §6 Parsing helpers
-- ─────────────────────────────────────────────────────────────────────

-- | Parse a whole hex literal, or 'Nothing' on any trailing junk.
parseHex :: String -> Maybe Int
parseHex s =
  case readHex s of
    [(n, "")] -> Just n
    _other    -> Nothing

-- | Strip leading and trailing whitespace.
trim :: String -> String
trim = dropWhileEnd isSpace . dropWhile isSpace

File: ports/python/src/unicode_python/security/identity/emoji_zwj_integrity.py
"""emoji-zwj-integrity — detection of malformed / unsanctioned emoji
ZWJ-sequence shapes per UTS #51 (the identity-layer detector I3).

Byte-faithful transliteration of the verified Rust reference
``ports/rust/src/security/identity/emoji_zwj_integrity.rs`` (itself a
transliteration of ``Unicode.Security.Identity.EmojiZwjIntegrity``).

Threat model. An adversary crafts an emoji-shaped codepoint sequence
containing one or more ``U+200D`` ZERO WIDTH JOINERs but violating the
sanctioned RGI ZWJ-sequence shape — by exceeding the RGI length cap, by
joining a non-emoji codepoint, by emitting adjacent ZWJ pairs, or by
overflowing the skin-tone count. Any non-RGI ZWJ-containing sequence is
renderer-dependent, and that renderer divergence is the attack surface.

Sanctioning data. UTS #51 defines the RGI ZWJ sequences in
``emoji-zwj-sequences.txt``, bundled in the port's own
``data/emoji-zwj-sequences.txt`` (never a host emoji library). The
registered set gives both the exact-match membership test
(:func:`is_registered_zwj_sequence`) and the ZWJ *alphabet* — every
distinct codepoint occurring at any position of any registered sequence,
excluding the joiner — which is the canonical "what may flank a ZWJ?"
predicate (:func:`is_emoji_target`).

Algorithm (one pass over ``input``).

    Phase 1 — collect ZWJ positions and the skin-tone count.
    Phase 2 — short-circuit ``Clear`` if there are no ZWJs and the
              skin-tone count is at most 1.
    Phase 3 — a registered RGI sequence is always ``Clear``.
    Phase 4 — check sub-threats by priority:
                1. ``DoubleZWJ``            ZWJ-ZWJ adjacency
                2. ``NonEmojiInjection``    ZWJ adjacent to a non-emoji cp
                3. ``OverLength``           sequence longer than the cap
                4. ``SkinToneOverflow``     skin-tone count >= 5
                5. ``UnregisteredSequence`` catch-all when ZWJs are present
                                            but the sequence is not registered.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Union

from ..calculus import ClassificationKind

# ─────────────────────────────────────────────────────────────────────
# §1 Constants
# ─────────────────────────────────────────────────────────────────────

# Conservative cap on the length of a sanctioned RGI ZWJ sequence
# (``maxRgiLength`` in the Lean spec). The longest current entry (a
# four-person family with skin tones) reaches ~13-14 codepoints; 16 is a
# safe upper bound.
MAX_RGI_LENGTH = 16

# The ZERO WIDTH JOINER codepoint.
ZWJ = 0x200D


# ─────────────────────────────────────────────────────────────────────
# §2 Sub-threat ADT + verdict
# ─────────────────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class DoubleZwj:
    """ZWJ-ZWJ adjacency; ``positions`` are the first ZWJ of each pair."""

    positions: list[int]


@dataclass(frozen=True, slots=True)
class NonEmojiInjection:
    """A ZWJ flanked by a non-emoji codepoint (or sitting at an input
    edge). ``non_emoji_cp`` is the offending codepoint (0 for an edge
    ZWJ)."""

    zwj_pos: int
    non_emoji_cp: int


@dataclass(frozen=True, slots=True)
class OverLength:
    """The sequence is longer than :data:`MAX_RGI_LENGTH`."""

    length: int
    max_length: int


@dataclass(frozen=True, slots=True)
class SkinToneOverflow:
    """Five or more skin-tone modifiers (the family-emoji maximum is
    four)."""

    count: int


@dataclass(frozen=True, slots=True)
class UnregisteredSequence:
    """ZWJs are present and no other sub-threat matched, but the sequence
    is not a registered RGI ZWJ sequence."""

    chain_len: int


SubThreat = Union[
    DoubleZwj,
    NonEmojiInjection,
    OverLength,
    SkinToneOverflow,
    UnregisteredSequence,
]


def sub_threat_tag(sub: SubThreat) -> str:
    """Fixture-row tag string for a sub-threat (matches ``SubThreat.tag``
    in the Lean/Rust reference)."""
    if isinstance(sub, DoubleZwj):
        return "DoubleZWJ"
    if isinstance(sub, NonEmojiInjection):
        return "NonEmojiInjection"
    if isinstance(sub, OverLength):
        return "OverLength"
    if isinstance(sub, SkinToneOverflow):
        return "SkinToneOverflow"
    if isinstance(sub, UnregisteredSequence):
        return "UnregisteredSequence"
    raise TypeError(f"sub_threat_tag: unknown SubThreat variant {sub!r}")


@dataclass(frozen=True, slots=True)
class Classification:
    """Top-level EmojiZwjIntegrity classification. ``is_clear``
    distinguishes the ``Clear`` variant from the ``Hazard`` variant; a
    hazard carries its sub-threat, the codepoint positions it implicates,
    and the decoded-byte projection (always empty here, kept for shape
    parity with the Lean ``Classification.hazard``)."""

    is_clear: bool
    sub: SubThreat | None = None
    positions: list[int] = field(default_factory=list)
    decoded: list[int] = field(default_factory=list)

    @property
    def tag(self) -> str | None:
        """Human-facing tag for a hazard, or ``None`` when clear."""
        if self.is_clear or self.sub is None:
            return None
        return sub_threat_tag(self.sub)

    @property
    def kind(self) -> ClassificationKind:
        """The classification kind (``CLEAR`` or ``HAZARD``)."""
        return (
            ClassificationKind.CLEAR
            if self.is_clear
            else ClassificationKind.HAZARD
        )


@dataclass(frozen=True, slots=True)
class Verdict:
    """The structured output of :func:`detect` (mirrors the Lean
    ``Verdict``)."""

    input: list[int]
    classify: Classification
    zwj_positions: list[int]
    chain_length: int
    is_registered_rgi: bool
    skin_tone_count: int


# ─────────────────────────────────────────────────────────────────────
# §3 RGI ZWJ-sequence data (bundled data/emoji-zwj-sequences.txt)
# ─────────────────────────────────────────────────────────────────────

_DATA_DIR = Path(__file__).resolve().parent.parent.parent / "data"


def _parse_zwj_sequences() -> list[list[int]]:
    """Parse the registered RGI ZWJ sequences from
    emoji-zwj-sequences.txt. Each non-comment row is
    ``<cp> <cp> ... ; RGI_Emoji_ZWJ_Sequence ; <desc> # <cmt>``; the
    codepoint list is the space-separated hex field before the first
    ``;``."""
    with (_DATA_DIR / "emoji-zwj-sequences.txt").open(
        "r", encoding="utf-8"
    ) as f:
        raw = f.read()
    out: list[list[int]] = []
    for raw_line in raw.splitlines():
        hash_idx = raw_line.find("#")
        body = raw_line if hash_idx < 0 else raw_line[:hash_idx]
        stripped = body.strip()
        if not stripped:
            continue
        seq_field = stripped.split(";")[0]
        seq: list[int] = []
        parsed_ok = True
        for token in seq_field.split():
            try:
                seq.append(int(token, 16))
            except ValueError:
                parsed_ok = False
                break
        if parsed_ok and seq:
            out.append(seq)
    return out


_ZWJ_SEQUENCES: list[list[int]] | None = None
_ZWJ_ALPHABET: frozenset[int] | None = None


def zwj_sequences() -> list[list[int]]:
    """Return the parsed registered RGI ZWJ sequences (cached after first
    call)."""
    global _ZWJ_SEQUENCES
    if _ZWJ_SEQUENCES is None:
        _ZWJ_SEQUENCES = _parse_zwj_sequences()
    return _ZWJ_SEQUENCES


def _build_zwj_alphabet() -> frozenset[int]:
    """The ZWJ alphabet: every distinct codepoint occurring at any
    position of any registered RGI ZWJ sequence, excluding the joiner
    U+200D itself."""
    out: set[int] = set()
    for seq in zwj_sequences():
        for cp in seq:
            if cp != ZWJ:
                out.add(cp)
    return frozenset(out)


def zwj_alphabet() -> frozenset[int]:
    """Return the ZWJ alphabet (cached after first call)."""
    global _ZWJ_ALPHABET
    if _ZWJ_ALPHABET is None:
        _ZWJ_ALPHABET = _build_zwj_alphabet()
    return _ZWJ_ALPHABET


def is_registered_zwj_sequence(cps: list[int]) -> bool:
    """True iff ``cps`` is exactly a registered RGI ZWJ sequence."""
    return any(seq == cps for seq in zwj_sequences())


def is_emoji_target(cp: int) -> bool:
    """True iff ``cp`` appears at some position of a registered RGI ZWJ
    sequence (the canonical "what may flank a ZWJ?" predicate)."""
    return cp in zwj_alphabet()


# ─────────────────────────────────────────────────────────────────────
# §4 Core predicates
# ─────────────────────────────────────────────────────────────────────


def is_zwj(cp: int) -> bool:
    """True iff ``cp`` is the ZWJ codepoint."""
    return cp == ZWJ


def is_emoji_modifier(cp: int) -> bool:
    """True iff ``cp`` is an emoji skin-tone modifier
    (U+1F3FB..U+1F3FF)."""
    return 0x1F3FB <= cp <= 0x1F3FF


def zwj_positions(input_cps: list[int]) -> list[int]:
    """Positions of every ZWJ in ``input``."""
    return [idx for idx, cp in enumerate(input_cps) if is_zwj(cp)]


def skin_tone_count(input_cps: list[int]) -> int:
    """Count of skin-tone modifier codepoints."""
    return sum(1 for cp in input_cps if is_emoji_modifier(cp))


def double_zwj_positions(input_cps: list[int]) -> list[int]:
    """Positions of the first ZWJ in each ZWJ-ZWJ adjacent pair."""
    out: list[int] = []
    for idx in range(len(input_cps)):
        if (
            idx + 1 < len(input_cps)
            and is_zwj(input_cps[idx])
            and is_zwj(input_cps[idx + 1])
        ):
            out.append(idx)
    return out


def first_non_emoji_injection(
    input_cps: list[int],
) -> tuple[int, int] | None:
    """The first ZWJ position where either neighbour is a non-emoji
    codepoint, as ``(zwj_pos, offending_cp)``. A ZWJ at an input edge (no
    preceding or no following codepoint) is itself an injection-class
    hazard, reported with offending codepoint 0."""
    for idx in range(len(input_cps)):
        if not is_zwj(input_cps[idx]):
            continue
        prev = None if idx == 0 else input_cps[idx - 1]
        nxt = input_cps[idx + 1] if idx + 1 < len(input_cps) else None
        if prev is not None and nxt is not None:
            if not is_emoji_target(prev):
                return (idx, prev)
            if not is_emoji_target(nxt):
                return (idx, nxt)
        elif prev is None:
            return (idx, 0)
        else:
            # prev present, nxt is None (trailing-edge ZWJ).
            return (idx, 0)
    return None


# ─────────────────────────────────────────────────────────────────────
# §5 Top-level detection
# ─────────────────────────────────────────────────────────────────────


def detect(input_cps: list[int]) -> Verdict:
    """The EmojiZwjIntegrity detection function."""
    zwjs = zwj_positions(input_cps)
    st_count = skin_tone_count(input_cps)
    is_rgi = is_registered_zwj_sequence(input_cps)
    chain_len = 0 if not zwjs else len(input_cps)

    if not zwjs and st_count <= 1:
        return Verdict(
            input=list(input_cps),
            classify=Classification(is_clear=True),
            zwj_positions=[],
            chain_length=0,
            is_registered_rgi=is_rgi,
            skin_tone_count=st_count,
        )

    if is_rgi:
        # Phase 3: a registered RGI sequence is always clear.
        classification = Classification(is_clear=True)
    else:
        # Phase 4.1: ZWJ-ZWJ adjacency.
        dzwj = double_zwj_positions(input_cps)
        if dzwj:
            classification = Classification(
                is_clear=False,
                sub=DoubleZwj(positions=list(dzwj)),
                positions=list(dzwj),
            )
        else:
            injection = first_non_emoji_injection(input_cps)
            if injection is not None:
                # Phase 4.2: ZWJ adjacent to a non-emoji codepoint.
                zwj_pos, offend_cp = injection
                classification = Classification(
                    is_clear=False,
                    sub=NonEmojiInjection(
                        zwj_pos=zwj_pos, non_emoji_cp=offend_cp
                    ),
                    positions=[zwj_pos],
                )
            elif len(input_cps) > MAX_RGI_LENGTH:
                # Phase 4.3: length cap.
                classification = Classification(
                    is_clear=False,
                    sub=OverLength(
                        length=len(input_cps), max_length=MAX_RGI_LENGTH
                    ),
                    positions=[],
                )
            elif st_count >= 5:
                # Phase 4.4: skin-tone overflow.
                classification = Classification(
                    is_clear=False,
                    sub=SkinToneOverflow(count=st_count),
                    positions=[],
                )
            elif zwjs:
                # Phase 4.5: catch-all for unregistered ZWJ sequences.
                classification = Classification(
                    is_clear=False,
                    sub=UnregisteredSequence(chain_len=len(input_cps)),
                    positions=list(zwjs),
                )
            else:
                classification = Classification(is_clear=True)

    return Verdict(
        input=list(input_cps),
        classify=classification,
        zwj_positions=list(zwjs),
        chain_length=chain_len,
        is_registered_rgi=is_rgi,
        skin_tone_count=st_count,
    )
